### PR TITLE
Core, Server, Addon-docs: Replace story indexers with new API

### PR DESCRIPTION
**Pull request by JReinhold**
Works on #23457 

## What I did

This PR does multiple things (shame on me. Sorry.)

### 1. Replaces all the internal _story_ indexes with the new API:

- CSF indexer
- Stories MDX indexer
- Server indexer

### 2. Handle `docsOnly` mode in Stories MDX

The new indexer handler now automatically removes story entries that are `docsOnly`\*, as defined by having the `"docsOnly"` tag. So the Stories MDX indexer now adds that tag when it sees `parameters.docsOnly`, which is actually set by `CsfFile.parse()` 🥵. 

\* `docsOnly` is a feature exclusive to Stories MDX, and happens whenever any `stories.mdx` files don't render any stories - so only Docs needs to appear in the sidebar. I assume this handling can be removed when we finally remove Stories MDX completely.

### 3. Minor improvements to the `server` renderer

- The default `stories` glob when initing a `server` project now includes `yaml` and `yml` extensions along with the existing `json` extensions. The indexer already handled this, it just wasn't in the globs
- The default stories for Page have now been converted to YAML to showcase/test this behavior
- The new indexer adds support for tags, and autodocs works mostly

Two aspects of autodocs that are broken in the server renderer:

- Any (iframed) stories beyond the first three always hangs forever in loading, consistently
- Controls don't work, they don't trigger any changes

Are these two points enough that I should remove it again, or is this better than nothing @shilman ?

## How to test

1. See Chromatic doesn't report any changes to stories (especially changing names, ids), as the new indexers should produce the same result.
2. See especially the unattached story at `?path=/docs/addons-docs-stories-mdx-unattached--docs`. It should be a lonely `unattached` Docs entry with no stories, because it's `docsOnly`.
3. For the server changes you have to create the sandbox manually, by checking out this repo and creating the `internal/server-webpack5` sandbox.

## Checklist

- [x] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "build", "documentation", "maintenance", "dependencies", "other"]`

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

### CODE REVIEW

File: code/renderers/server/src/preset.ts [hunk 1]
```diff
@@ -1,36 +1,30 @@
...
-    },
-    ...(existingIndexers || []),
-  ];
+type FileContent = {
+  title: ComponentTitle;
+  tags?: Tag[];
+  stories: { name: StoryName; tags?: Tag[] }[];
 };
+
+export const indexers: StorybookConfig['indexers'] = (existingIndexers) => [
```
Comment (ndelangen) [inline code/renderers/server/src/preset.ts:11]: So this changed from `storyIndexers` to `indexers`, right.
Maybe I'm missing it in this PR, but I'd expect this change to be actually used somewhere, was that already done in another PR?

Might this be marked as experimental in any way?

File: code/addons/docs/src/preset.ts [hunk 2]
```diff
@@ -129,23 +129,36 @@ async function webpack(
...
+      const { compile } = global.FEATURES?.legacyMdx1
+        ? await import('@storybook/mdx1-csf')
+        : await import('@storybook/mdx2-csf');
+      code = await compile(code, {});
+      const csf = loadCsf(code, { ...opts, fileName }).parse();
+
+      // eslint-disable-next-line no-underscore-dangle
+      return Object.entries(csf._stories).map(([exportName, story]) => {
+        const docsOnly = story.parameters?.docsOnly;
+        const tags = (story.tags ?? csf.meta.tags ?? []).concat(docsOnly ? 'docsOnly' : []);
```
Comment (tmeasday) [inline code/addons/docs/src/preset.ts:146]: Do these stories end up with the `stories-mdx` tag still?

File: code/lib/core-server/src/utils/StoryIndexGenerator.test.ts [hunk 3]
```diff
@@ -42,16 +42,20 @@ const storiesMdxIndexer: Indexer = {
...
+        type: 'story',
+        importPath: fileName,
+        exportName,
+        name: story.name,
+        title: csf.meta.title,
+        metaId: csf.meta.id,
+        tags,
+        __id: story.id,
+      };
+    });
```
Comment (tmeasday) [inline code/lib/core-server/src/utils/StoryIndexGenerator.test.ts:58]: Should we use the real thing now?

File: code/lib/core-server/src/utils/StoryIndexGenerator.ts [hunk 4]
```diff
@@ -346,8 +346,12 @@ export class StoryIndexGenerator {
       });
     }
 
+    const entriesWithoutDocsOnlyStories = entries.filter(
+      (entry) => !(entry.type === 'story' && entry.tags.includes('docsOnly'))
```
Comment (tmeasday) [inline code/lib/core-server/src/utils/StoryIndexGenerator.ts:350]: Maybe this should be something more obscure like `stories-mdx-docsOnly`? Seems unfortunately to reserve a tags meaning for this deprecated reason.

**Review by tmeasday — APPROVED** [review]
Looks good but I have some questions.

File: code/addons/docs/src/preset.ts (hunk 2 above)
Comment (JReinhold) [reply code/addons/docs/src/preset.ts:146]: yes, [`@storybook/mdx2-csf` takes care of that](https://github.com/storybookjs/mdx2-csf/blob/2fa8864e919bec8f2d15f7a5eed715e64212f29c/src/sb-mdx-plugin.ts#L320C1-L320C1).

Although if tags are set on a story, it would overwrite it. It has always been like that, should I fix it?

Comment (tmeasday) [reply code/addons/docs/src/preset.ts]: I don't know if you can set tags in `.stories.mdx` files, but maybe you should to be safe?

File: docs/snippets/common/storybook-main-csf-indexer.ts.mdx [hunk 5]
```diff
@@ -5,10 +5,21 @@ import { readFileSync } from 'fs';
...
+        exportName,
+        name: story.name,
+        title: csf.meta.title,
+        metaId: csf.meta.id,
+        tags: story.tags ?? csf.meta.tags,
+        __id: story.id,
+      }));
     };
 
     return [
```
Comment (kylegach) [inline docs/snippets/common/storybook-main-csf-indexer.ts.mdx:25]: @JReinhold — I think `indexer` should be `index` here, to match the update to the TS snippet?

**Review by ndelangen — APPROVED** [review]

File: code/lib/core-server/src/presets/common-preset.ts [hunk 6]
```diff
@@ -194,20 +194,31 @@ export const features = async (
...
-    },
-    ...(existingIndexers || []),
-  ];
+export const csfIndexer: Indexer = {
+  test: /\.stories\.(m?js|ts)x?$/,
+  index: async (fileName, options) => {
+    const csf = (await readCsf(fileName, options)).parse();
+
+    // eslint-disable-next-line no-underscore-dangle
+    return Object.entries(csf._stories).map(([exportName, story]) => {
```
Comment (shilman) [inline code/lib/core-server/src/presets/common-preset.ts:203]: Can we add this to a getter on `CsfFile`? The code seems to be duplicated all over the place and that would make this PR a lot more DRY.

**Review by shilman — COMMENTED** [review]
Looks great! One suggestion though

File: code/lib/core-server/src/presets/common-preset.ts (hunk 6 above)
Comment (shilman) [reply code/lib/core-server/src/presets/common-preset.ts:203]: `CsfFile.indexEntries`?

Comment (JReinhold) [reply code/lib/core-server/src/presets/common-preset.ts:203]: Yeah I've been thinking something similar, I'll experiment with a getter for  `indexInputs` and stories with `exportName` so you don't need the funky use of `_stories`.

**Review by shilman — APPROVED** [review]
